library/modules/Gui.cpp

```diff
@@ -1574,6 +1707,239 @@ void Gui::showAutoAnnouncement(
     addCombatReportAuto(unit2, flags, id);
 }
 
+int Gui::autoDFAnnouncement(df::report_init r, string message)
+{   // Reverse-engineered from DF announcement code
+
+    if (!world->unk_26a9a8) // TODO: world->show_announcements
+        return 1;
+
+    df::announcement_flags a_flags;
+    if (is_valid_enum_item(r.type))
+        a_flags = df::global::d_init->announcements.flags[r.type];
+    else
+        return 2;
```

Comment:
Can these return codes be made an enum instead of magic numbers? Such an enum should probably be defined in `Gui.h` because `LuaApi.cpp` also needs it, and any callers would benefit from it too.

I wouldn't worry too much about exposing the enum to Lua yet, but if you want to, you could put it in df-structures as `report_failure_reason` or something along those lines. If it's an enum that DF doesn't use (I'm not sure if you added these codes yourself or if they were part of the reverse-engineered code), I would follow our existing pattern of prefixing it with `dfhack_`.

Reply:
I added them so dfhack scripters could know why their announcements aren't displaying. ``debugfilter`` can handle this, and we can return boolean if the announcement printed or not.

The remaining use case is that somebody wants to do something different based on why it failed, which is niche enough that they can write their own function.

Reply:
I know that's what the numbers are for - my point is that they ought to be in an enum instead of just being hardcoded like this. If you change `2` here, you have to change it in at least one other place. (I misinterpreted what was going on in LuaApi.cpp earlier - looks like this is only really used in Gui.cpp at the moment, so the enum could be internal to that file.)

Also, in terms of API design, I really think the numbers should be possible to interpret from code. Right now, there is an overload of `autoDFAnnouncement()` that displays an announcement, optionally prints an error message, and returns. I think the responsibility of printing an error message should be left to the caller, which knows where it should go. Maybe the caller wants to print it to a file, or display it to the screen, or not print it immediately. `debugfilter` can't help with any of those use-cases, and we shouldn't have to worry about supporting all of those use-cases in every function that can fail with an error.

Reply:
If the debug messages are output inline (where the error occurs) instead of here, there may not be a need for an enum at all.

Reply:
Sure, I think that's also fine. I guess I was thinking of them as more "errors" than "diagnostics". If they're not deemed important enough to pass up to the caller, that's fine by me - but the status codes _are_ being passed to the caller currently (even to Lua), which is why I thought they might be important.

So I guess to summarize my opinions here:
- If the status codes are returned to the caller, translation to messages should be done by the caller, if desired
- If the status codes aren't returned, I'd be fine with printing messages inline like you said
- If status codes appear more than once, they ought to be an enum (at least within Gui.cpp) to improve readability